### DISCUSSION TITLE
Derive level-of-detail / macro structure graph

### DISCUSSION THREAD

**Discussion by jpmaterial** [networkx/networkx#4908] [Q&A]
I have a graph embedded in 3D with lots of nodes (derived from an STL). I would like to simplify the graph to its macro structure / reduce its level of detail. Either introducing new vertices or reusing a subset of the present ones would be okay. The derived graph's edges should be the center of the areas they simplify like a skeleton/medial axis graph.

I'm not sure how to rigorously define it, maybe like so: Given a graph G with N nodes, find a subgraph H Pareto optimal in M,d where M is the number of nodes in the H and d is the longest weighted distance from a node in G to a node in H.

This is just one possible definition. Maybe another one is possible which defines a unique derived graph.

Things I have tried or considered:

- Simplify the spanning tree -> not tried yet, no idea how to prune
- Iteratively contract the cycles (plus keeping nodes of degree 1) -> does not seem to work

The below gives an idea of what I'm working with and aiming for:
![image](https://user-images.githubusercontent.com/15215287/122420543-0e774280-cf8c-11eb-947b-8135bbd007c8.png)

**Comment by rossbar** [comment]
I have no idea whether this fits the bill for your application, but there's a new [dedensification function](https://networkx.org/documentation/latest/reference/algorithms/generated/networkx.algorithms.summarization.dedensify.html#networkx-algorithms-summarization-dedensify) that will be available in the upcoming release (you can try it now with the networkx 2.6 release candidate, e.g. `pip install --pre networkx`).

Another simple option would be to create a regularly-gridded space and simply threshold the grid based on the number of nodes in each cell --- `np.histogram2d` seems like it would work nicely for the example image above, assuming that you have the absolute 2d positions for all the nodes stored in the graph.

**Reply by jpmaterial** [reply]
Thanks for your suggestions. I will try the dendensification as soon as 2.6 is released.

Another approach I came up with is working with what could be called "depth". (Above I didn't explicitly mention that the depicted graph is already a subgraph of a graph based on a triangle mesh.) Mark the border nodes (those who in the full graph have neighbors not in the subgraph) with depth 0. Then propagate through the graph: If a node has neighbors with defined depth, set its depth to the minimum of its neighbors' depths + 1. When all nodes have a depth value, only keep nodes with locally maximal depth. Then connect these maxima preferring nodes with higher depth.

In that approach I see two tasks that networkx could potentially support: Propagating the depths values (is there e.g. a way to get all neighbors to nodes with a given 'depth' value?) and the path finding based on depth values. Does networkx have any handy algorithms here?

**Reply by dschult** [reply]
The function `nx.shortest_path_length` gives the depth values you are talking about, right? If not, it should be easy to use the [depth-first-search iterator](https://networkx.org/documentation/stable/reference/algorithms/traversal.html) `dfs_edges` to build what you need.
